src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationAssembliesProvider.cs

```diff
@@ -0,0 +1,256 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    internal class ApplicationAssembliesProvider
+    {
+        internal static HashSet<string> ReferenceAssemblies { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "Microsoft.AspNetCore.Mvc",
+            "Microsoft.AspNetCore.Mvc.Abstractions",
+            "Microsoft.AspNetCore.Mvc.ApiExplorer",
+            "Microsoft.AspNetCore.Mvc.Core",
+            "Microsoft.AspNetCore.Mvc.Cors",
+            "Microsoft.AspNetCore.Mvc.DataAnnotations",
+            "Microsoft.AspNetCore.Mvc.Formatters.Json",
+            "Microsoft.AspNetCore.Mvc.Formatters.Xml",
+            "Microsoft.AspNetCore.Mvc.Localization",
+            "Microsoft.AspNetCore.Mvc.Razor",
+            "Microsoft.AspNetCore.Mvc.Razor.Extensions",
+            "Microsoft.AspNetCore.Mvc.RazorPages",
+            "Microsoft.AspNetCore.Mvc.TagHelpers",
+            "Microsoft.AspNetCore.Mvc.ViewFeatures",
+        };
+
+        /// <summary>
+        /// Returns an ordered list of application assemblies.
+        /// <para>
+        /// The order is as follows:
+        /// * Entry assembly
+        /// * Assemblies specified in the application's deps file ordered by name.
+        /// <para>
+        /// Each assembly is immediately followed by assemblies specified by annotated <see cref="RelatedAssemblyAttribute"/> ordered by name.
+        /// </para>
+        /// </para>
+        /// </summary>
+        public IEnumerable<Assembly> ResolveAssemblies(Assembly entryAssembly)
+        {
+            var dependencyContext = LoadDependencyContext(entryAssembly);
+
+            IEnumerable<AssemblyItem> assemblyItems;
+
+            if (dependencyContext == null)
+            {
+                assemblyItems = new[] { GetAssemblyItem(entryAssembly) };
+            }
+            else
+            {
+                assemblyItems = ResolveFromDependencyContext(dependencyContext);
+            }
+
+            assemblyItems = assemblyItems
+                .OrderBy(item => item.Assembly == entryAssembly ? 0 : 1)
+                .ThenBy(item => item.Assembly.FullName, StringComparer.Ordinal);
+
+            foreach (var item in assemblyItems)
+            {
+                yield return item.Assembly;
+
+                foreach (var associatedAssembly in item.RelatedAssemblies.Distinct().OrderBy(assembly => assembly.FullName, StringComparer.Ordinal))
+                {
+                    yield return associatedAssembly;
+                }
+            }
+        }
+
+        protected virtual DependencyContext LoadDependencyContext(Assembly assembly) => DependencyContext.Load(assembly);
```

Comment:
Why does this need to be protected virtual (on an internal class)

Reply:
We use it in tests to mock up a DependencyContext